### ISSUE TITLE
Possible mem leak in tf.random.categorical

### ISSUE THREAD

**Issue by feidieufo** [tensorflow/tensorflow#36034]
<em>Please make sure that this is a bug. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow):Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):Ubuntu 16.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:No
- TensorFlow installed from (source or binary):binary
- TensorFlow version (use command below):2.0.0
- Python version:3.6.9
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version:10.0.130
- GPU model and memory:1060  6G

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with: 1. TF 1.0: `python -c "import
tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"` 2. TF 2.0: `python -c
"import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**
If I set seed, the memory keeps increasing between loops.But if I remove seed, the memory will remains the same.

**Describe the expected behavior**
Whether I set random seed or not，the memory will remains the same.

**Code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate the problem.
```python
import tensorflow as tf
import numpy as np
import tensorflow_probability as tfp
import psutil
import os

if __name__ == '__main__':
    act_dim = 8

    model = tf.keras.Sequential([tf.keras.layers.Input(500, ),
                                 tf.keras.layers.Dense(500, activation="relu"),
                                 tf.keras.layers.Dense(act_dim, activation="softmax")])

    tf.random.set_seed(0)
    print("memory used:", psutil.Process(os.getpid()).memory_info().rss)

    for j in range(100):
        for i in range(1000):
            x = np.random.rand(500, 500)
            prob = model(x)

            dist = tfp.distributions.Categorical(probs=prob)
            a = dist.sample()
        print("memory used:", psutil.Process(os.getpid()).memory_info().rss)
```

**Other info / logs**
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached.

**Comment by gadagashwini** [comment]
Issue is replicating on colab with Tf 2.0. 
Please find the gist [here](https://colab.sandbox.google.com/drive/17HhoZooFxxS4iukJAsJDZ6s1DiKGz7IX?authuser=1#scrollTo=kuiNQi0c9EbB). Thanks!

**Comment by csuter** [comment]
Thanks for reporting this. I created [this notebook](https://colab.research.google.com/drive/1gsT11Gf7hQVSDplsSEswTJrMggX_GHt4) which demonstrates the issue using `tf.random.categorical` directly, without any invocation of TensorFlow Probability library code. This suggests the issue is at the TF level, not TFP. Kicking back to Yasir to reassign appropriately on the TF side. Please let me know if I can be of further assistance!

**Comment by csuter** [comment]
Here's the code from that notebook, which is not viewable outside Google:

I ran this with `tf-nightly` package.

```python
import tensorflow.compat.v2 as tf
tf.enable_v2_behavior()
import numpy as np
import psutil
import os

if __name__ == '__main__':
    act_dim = 8

    model = tf.keras.Sequential([tf.keras.layers.Input(500, ),
                                 tf.keras.layers.Dense(500, activation="relu"),
                                 tf.keras.layers.Dense(act_dim, activation="softmax")])

    tf.random.set_seed(0)
    print("memory used:", psutil.Process(os.getpid()).memory_info().rss)

    for j in range(100):
        for i in range(1000):
            x = np.random.rand(500, 500).astype(np.float32)
            prob = model(x)

            a = tf.random.categorical(logits=tf.math.log(prob), num_samples=1)
        print("memory used:", psutil.Process(os.getpid()).memory_info().rss)
```

**Comment by feidieufo** [comment]
> Here's the code from that notebook, which is not viewable outside Google:
> 
> I ran this with `tf-nightly` package.
> 
> ```python
> import tensorflow.compat.v2 as tf
> tf.enable_v2_behavior()
> import numpy as np
> import psutil
> import os
> 
> if __name__ == '__main__':
>     act_dim = 8
> 
>     model = tf.keras.Sequential([tf.keras.layers.Input(500, ),
>                                  tf.keras.layers.Dense(500, activation="relu"),
>                                  tf.keras.layers.Dense(act_dim, activation="softmax")])
> 
>     tf.random.set_seed(0)
>     print("memory used:", psutil.Process(os.getpid()).memory_info().rss)
> 
>     for j in range(100):
>         for i in range(1000):
>             x = np.random.rand(500, 500).astype(np.float32)
>             prob = model(x)
> 
>             a = tf.random.categorical(logits=tf.math.log(prob), num_samples=1)
>         print("memory used:", psutil.Process(os.getpid()).memory_info().rss)
> ```

Thank you for your reply.I ran this code in my environment(tensorflow-gpu-2.0.0 by using conda install tensorflow-gpu),If I set seed,the the memory keeps increasing,if I do not set seed,the memory will keep remains the same.But when I ran this code in colab,whether I set random seed or not，the memory will keep increasing.I don't know why.

**Comment by wangpengmit** [comment]
I don't know why `tf.random.set_seed` makes a difference here, but `tf.random.categorical` has known memory usage issues, because it uses the [Gambel-distribution trick](https://lips.cs.princeton.edu/the-gumbel-max-trick-for-discrete-distributions/) which uses a lot of memory. You can try this code that manually implement the sampler:

```python
def categorical(logits, num_samples, dtype=None, seed=None):
  logits = tf.convert_to_tensor(logits, name="logits")
  seed1, seed2 = tf.compat.v1.get_seed(seed)

  dt = tf.float32
  batch = logits.shape.dims[0].value
  num_classes = logits.shape.dims[1].value
  if num_classes == 0:
    # Delegates to native op to raise the proper error.
    return tf.random.categorical(
        logits, num_samples, seed=seed1, seed2=seed2, output_dtype=dtype)
  # u ~ Uniform[0.0, 1.0)
  u = tf.random.uniform(
      shape=[batch, num_samples],
      seed=seed1, seed2=seed2, dtype=dt)
  # for numerical stability
  max_logit = tf.math.reduce_max(logits, axis=1, keepdims=True)
  logits = logits - max_logit
  pdf = tf.cast(tf.math.exp(logits), dtype=dt)  # not normalized
  cdf = tf.math.cumsum(pdf, axis=1)
  cdf_last= cdf[:, -1:]
  u = u * cdf_last
  if num_samples == 0 or batch == 0:
    # A tf.searchsorted bug workaround
    return tf.zeros([batch, num_samples])
  else:
    return tf.searchsorted(cdf, u, side="right")  # upper_bound
```

**Comment by IanRDavies** [comment]
I'm having a similar issue for some Reinforcement Learning code where the issue is more pronounced. If I set a seed and then train an agent which samples actions randomly at each time step the memory leak becomes very pronounced.

ystem information

    Have I written custom code (as opposed to using a stock example script provided in TensorFlow):Yes
    OS Platform and Distribution (e.g., Linux Ubuntu 16.04):Ubuntu 18.04
    Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:No
    TensorFlow installed from (source or binary):binary
    TensorFlow version (use command below):2.0.0 (also verified on 2.1.0)
    Python version:3.7.5
    Bazel version (if compiling from source): N/A
    GCC/Compiler version (if compiling from source): N/A
    CUDA/cuDNN version: N/A
    GPU model and memory: N/A

**Comment by foxik** [comment]
I believe this may be caused by caching which `convert_to_eager_tensor` performs. Notably look at https://github.com/tensorflow/tensorflow/blob/7072568ed6b735e347fb87bc84a2b83daf806e3f/tensorflow/python/framework/constant_op.py#L68-L72 which mentions that some values can be cached.

In the C++ code, https://github.com/tensorflow/tensorflow/blob/7072568ed6b735e347fb87bc84a2b83daf806e3f/tensorflow/python/eager/pywrap_tensor.cc#L307-L313 performs the caching -- currently for Python scalars only.

When a seed is set, it causes operation seeds to be generated by random.Random generator, which are probably then converted to eager tensors -- and the problem arises.

**Comment by tilakrayal** [comment]
@feidieufo ,
Can you please take a look at this [comment](https://github.com/tensorflow/tensorflow/issues/36034#issuecomment-576930993) and try in latest stable version v2.7 and let us know if the issue still persists.Thanks
